### ISSUE TITLE
NPE posterior exceeding prior bounds and failure for parameter prediction in simple linear model

### ISSUE THREAD

**Issue by Aoqige** [maho3/ltu-ili#203]
### Description
I'm using ltu-ili to perform SBI using the Neural Posterior Estimation (NPE) method. My model is a simple linear equation:

$y = θ_{0} * x + θ_{1}.$, where $θ_{0}$ and $θ_{1}$ are the parameters to be inferred.

### Configuration

- Method: NPE
- ltd-ili version: v 0.1.0
- Prior: $θ_{0}, \, θ_{1}$ ∼Uniform[0,1]
- Noise: None (deterministic simulator for testing purposes)
- Data: y generated from the linear model above.

Observed Behavior: I have encountered two unexpected results during my testing (plots attached below):

1. Posterior Beyond Prior Bounds: The marginal posterior for θ1 clearly extends beyond the [0,1] range defined in the prior. This suggests that the prior constraints may be not strictly enforced or respected by the density estimator.

<img width="489" height="490" alt="Image" src="https://github.com/user-attachments/assets/c7e3e561-1e4d-468f-aebe-b960c8cde296" />

2. Bad behavior for θ1: The model can recover θ0 with high accuracy, but the prediction for θ1 appears to be bad and "noisy" showing almost no correlation with the true values.

<img width="692" height="370" alt="Image" src="https://github.com/user-attachments/assets/6b908ad7-0076-4fc7-b401-372132f49cbc" />

### Questions:

Is it expected for NPE in ltu-ili to produce posterior samples outside the prior support, or is there a specific configuration (e.g., a transform or boundary handling) I should apply?

Regarding the second issue, does the failure of the "Predicted vs True" plot for θ1 is a normal phenomenon, or is this a known limitation when parameters are highly correlated in NPE?

concrete code is in the following page.

**Comment by Aoqige** [comment]
%load_ext autoreload
%autoreload 2

import warnings
warnings.filterwarnings('ignore')

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import torch
import torch.nn as nn

import ili
from ili.dataloaders import NumpyLoader
from ili.inference import InferenceRunner
from ili.validation.metrics import PosteriorCoverage, PlotSinglePosterior

device = 'cuda' if torch.cuda.is_available() else 'cpu'

# create synthetic catalog
def simulator(params):
    # create toy simulations
    x = np.arange(10) + 10
    y = params[0] * x +  params[1] 
    # y += np.random.randn(len(x)) * 0.3
    return y

seed_sim = 12345
np.random.seed(seed_sim)
theta = np.random.rand(2000, 2)  # 200 simulations, 3 parameters
x = np.array([simulator(t) for t in theta])

# make a dataloader
loader = NumpyLoader(x=x, theta=theta)

# define a prior 
prior = ili.utils.Uniform(low=[0, 0], high=[1, 1], device=device)

# instantiate your neural networks to be used as an ensemble
repeats_maf = 2 # not to have to duplicate code for a large ensemble of identical architectures
nets = [ili.utils.load_nde_sbi(engine='NPE', model='maf', hidden_features=100, num_transforms=5, repeats = repeats_maf)]
# nets = [ili.utils.load_nde_sbi(engine='NLE', model='maf', hidden_features=150, num_transforms=20)]

# define training arguments
train_args = {
    'training_batch_size': 800,
    'learning_rate': 2e-5
}

# initialize the trainer
runner = InferenceRunner.load(
    backend='sbi',
    engine='NPE',
    prior=prior,
    nets=nets,
    device=device,
    train_args=train_args,
    proposal=None,
    out_dir=None
)

# train the model
posterior_ensemble, summaries = runner(loader=loader)

# use ltu-ili's built-in validation metrics to plot the posterior for this point
theta_test = np.random.rand(1, 2) 
x_test = np.array([simulator(t) for t in theta_test])

metric = PlotSinglePosterior(
    num_samples=1000, sample_method='direct', 
    labels=[f'$\\theta_{i}$' for i in range(2)]
)
fig = metric(
    posterior=posterior_ensemble,
    x_obs = x_test[0], theta_fid=theta_test[0],
    plot_kws=dict(fill=True)
)

# Drawing samples from the ensemble posterior
metric = PosteriorCoverage(
    num_samples=1000, sample_method='direct', 
    labels=[f'$\\theta_{i}$' for i in range(2)],
    plot_list = ["coverage", "histogram", "predictions", "tarp"],
    out_dir=None
)

fig = metric(
    posterior=posterior_ensemble, # NeuralPosteriorEnsemble instance from sbi package
    x=x[::10], theta=theta[::10]
)

**Comment by Aoqige** [comment]
Sorry, I realized these were not bugs of the library but due to my own misunderstanding.

Regarding the first issue, I noticed that the prior is not necessarily strictly enforced (log_prob = -inf outside the prior bounds). For the second issue, the poor performance of the second parameter was related to the number of data points. After increasing the number of data points to 30, the issue was resolved and the predictions became accurate.

Closing this now.

**Comment by maho3** [comment]
Hi @Aoqige , thanks for your comments.

Yes, so your resolutions are correct.

1. The ltu-ili code actually does restrict your samples to the prior support (in your case [0,1]). What it does is the flows generate samples in [-inf,inf] , and then it checks whether each sample is within prior support before returning. However, the reason your first plot shows some samples below 0 is because the Seaborn corner plot uses a KDE to draw the contours and density plots. The KDE supposes Gaussian blobs at each sample point to create a smooth distribution. The plot generation is not restricted to the prior range, but the raw samples are.
2. Yes, that is likely what happened. The model likely didn't learn to break the degeneracy with few training points.

I hope this helps ! I'll close this issue now, but feel free to open it again if there are more questions.
